crates/task-impls/src/consensus.rs

```diff
@@ -870,24 +856,25 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
                     ));
             }
             HotShotEvent::QuorumProposalValidated(proposal) => {
-                let consensus = self.consensus.upgradable_read().await;
+                let consensus_reader = self.consensus.upgradable_read().await;
+                let mut consensus_writer = self.consensus.write().await;
```

Comment:
We'll always have to wait for the write lock anyway, couldn't we just take the single write lock for the entire thing?

Reply:
This one's tricky because of the mutability of the borrow for the callback. The `visit_leaf_ancestors` method requires an immutable borrow. I was thinking we could maybe change the signature, but it seems like that creates the need to change a couple spots (and add some more contention for write locks). But I'd be open to ideas.

Reply:
Oh, I see! That makes sense, I think this is fine as-is then

Reply:
Actually, looking at this again -- I think this is what causes the deadlock you were getting (the `async_block_on` in the predicates is probably fine, or at the very least not the sole issue). 

The write lock can't be taken until the read lock is dropped, so this line will just await forever. I think the only way to actually do this is to create a temporary `let mut decided_upgrade_cert = None`, then conditionally update `consensus.decided_upgrade_cert` right after the read lock gets upgraded